### DISCUSSION TITLE
Can I programmatically style text in a TextArea widget

### DISCUSSION THREAD

**Discussion by SZRabinowitz** [Textualize/textual#5336] [General]
is it possible for me to programmatically color certain characters inside a textarea, not syntax highlighting.

**Comment by SZRabinowitz** [comment]
Okay, for anyone with my problem, here is a solution. 
```python
textarea = TextArea()
textarea._highlights[line_number].append((start_collumn, end_collumn, highlight_name))
```


Format of highlight_name
highlight_name is a string that corresponds to a specific syntax element in the language being parsed. Examples might include:
"keyword"
"string"
"comment"
"function.name"
"type.identifier"
These names depend on the conventions used in the Tree-sitter grammar and highlight query for the language.



you can customize the color of the highlights by making a [custom theme](https://textual.textualize.io/widgets/text_area/#custom-themes). 




BTW: This is a big value of open source. I was able to dig through the source code and see how highlighting is implemented under the hood. Thank you textualize.


Also BTW: This is not a supported feature, so it may change without notice.

**Comment by SZRabinowitz** [comment]
Okay here is a custom TextArea widget allowing you to highlight arbitrary selections of text with any color from the rich library

```python
import sys

from rich._palettes import EIGHT_BIT_PALETTE, STANDARD_PALETTE, WINDOWS_PALETTE
from rich.color import ANSI_COLOR_NAMES, Color
from rich.style import Style
from textual.widgets import TextArea

IS_WINDOWS = sys.platform == "win32"


class TextAreaWithHighlights(TextArea):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        rich_colors = sorted((v, k) for k, v in ANSI_COLOR_NAMES.items())

        for color_number, name in rich_colors:
            palette = (
                WINDOWS_PALETTE
                if IS_WINDOWS and color_number < 16
                else STANDARD_PALETTE
                if color_number < 16
                else EIGHT_BIT_PALETTE
            )
            color = palette[color_number]
            self._theme.syntax_styles[name] = Style(color=Color.from_rgb(*color))

    def highlight(
        self, row: int, start_column: int, end_column: int, color: str
    ) -> None:
        self._highlights[row].append((start_column, end_column, color))
```

to highlight a color: `textarea.highlight(row_number, start_collumn, end_collumn, color)`
for example: `textarea.highlight(0, 3, 7, "red")`

the color can be any color in [this list](https://github.com/Textualize/rich/blob/master/rich/color.py#L49). 

here is an example:
```python
class MyApp(App):
    def compose(self) -> ComposeResult:
        some_text = """here is some text
green blue red yellow
here is more text
here is more text green blah blah
red yellow blue gree orange purple bright_red magenta
"""
        yield TextAreaWithHighlights(some_text)

    def on_mount(self) -> None:
        colors = [
            "red",
            "green",
            "blue",
            "yellow",
            "orange",
            "purple",
            "magenta",
            "bright_red",
        ]
        textarea: TextAreaWithHighlights = self.query_one(TextAreaWithHighlights)
        text = textarea.text
        for row, line in enumerate(text.splitlines()):
            for col in colors:
                if col in line:
                    index = line.find(col)
                    textarea.highlight(row, index, index + len(col), col)


if __name__ == "__main__":
    app = MyApp()
    app.run()
```

**Reply by SZRabinowitz** [reply]
@willmcgugan Do you have any interest in adding this ability to the native textual textarea?

**Reply by SZRabinowitz** [reply]
Good news is this worked. I used it for an [advent of code solution visualization](https://youtu.be/q0sGU2Ya8NQ)

**Reply by TomJGooding** [reply]
@SZRabinowitz Nice visualization, certainly useful for that particular puzzle!
